File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def dfs(self, root):
        if root == None: return [None, 0]

        left = self.dfs(root.left)
        right = self.dfs(root.right)

        if left[1] > right[1]: return [left[0], left[1] + 1]
        if right[1] > left[1]: return [right[0], right[1] + 1]

        return [root, left[1] + 1]
 
    def lcaDeepestLeaves(self, root):
        return self.dfs(root)[0]
        

File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
record Pair(TreeNode node, int depth) {}

class Solution {
    
    private Pair dfs(TreeNode root) {
        if(root == null) return new Pair(null, 0);

        Pair left = dfs(root.left);
        Pair right = dfs(root.right);

        if(left.depth() > right.depth()) return new Pair(left.node(), left.depth() + 1);
        if(right.depth() > left.depth()) return new Pair(right.node(), right.depth() + 1);

        return new Pair(root, left.depth() + 1);
    }

    public TreeNode lcaDeepestLeaves(TreeNode root) {
        Pair p = dfs(root);
        return p.node();
    }
}